### PR TITLE
Add  GlobalPhase to all devices

### PR DESCRIPTION
**Pull request by vincentmr**
### Before submitting

Please complete the following checklist when submitting a PR:

- [x] All new features must include a unit test.
      If you've fixed a bug or added code that should be tested, add a test to the
      [`tests`](../tests) directory!

- [x] All new functions and code must be clearly commented and documented.
      If you do make documentation changes, make sure that the docs build and
      render correctly by running `make docs`.

- [x] Ensure that the test suite passes, by running `make test`.

- [x] Add a new entry to the `.github/CHANGELOG.md` file, summarizing the
      change, and including a link back to the PR.

- [x] Ensure that code is properly formatted by running `make format`. 

When all the above are checked, delete everything above the dashed
line and fill in the pull request template.

------------------------------------------------------------------------------------------------------------

**Context:**
Prior to #516 , the Lightning devices could not properly run Grover's algorithm (among others) because `GlobalPhase` and `C(GlobalPhase)` are not supported. This would also be true of any circuit which ends up with `GlobalPhase` or `C(GlobalPhase)` after decomposition. In addition, L-GPU's custom gate cache was not passed a parameter to distinguish between gates. This meant, for example, that `MultiControlledX([0, 1, 2], "111")` and `MultiControlledX([0, 2], "00")` were applied as the same operation. This likely affected any circuit with more than one version of 
```
"QubitUnitary",
"ControlledQubitUnitary",
"MultiControlledX",
"DiagonalQubitUnitary",
"PSWAP",
"OrbitalRotation",
```

**Description of the Change:**
Add `GlobalPhase` and `C(GlobalPhase)` support in L-GPU and L-Kokkos. For the sake of time, `C(GlobalPhase)` is implemented by generating the diagonal entries with `global_phase_diagonal` at the Python layer, and then passing it through the gate matrix argument down to the C++ layer and adding a special branch for it in `applyOperation`. To fix the gate cache, we pass the gate hash as a parameter when applying any of the above gates to make sure the gate cache adds a new matrix. 

**Benefits:**
Bug fix, can now run Grover with L-GPU and L-Kokkos.

**Possible Drawbacks:**
Higher memory demand in the gate cache.

**Related GitHub Issues:**

### CODE REVIEW

**Comment by vincentmr** [comment]
Grover algorithm returns wrong results with [sc-51671]

File: pennylane_lightning/core/src/simulators/lightning_gpu/initSV.cu
Comment (mlxd) [inline pennylane_lightning/core/src/simulators/lightning_gpu/initSV.cu:124]: This kind of branching may be inefficient for CUDA kernels. Are we sure the `num_sv` value is at least as wide as a CUDA warp so that each SM is run at full capacity?

File: pennylane_lightning/core/src/simulators/lightning_gpu/initSV.cu
Comment (mlxd) [inline pennylane_lightning/core/src/simulators/lightning_gpu/initSV.cu:144]: Since these values are usually inferred from physical limits (as in, the total number of runnable threads, blocks, etc), are we sure the chosen numbers here, and arguments to the functions are sufficient to fit everything for all use-cases using this 1D indexing? Will we need to consider 2/3D blocks if things get larger?

File: pennylane_lightning/core/src/simulators/lightning_gpu/initSV.cu
Comment (mlxd) [inline pennylane_lightning/core/src/simulators/lightning_gpu/initSV.cu:166]: Same here as earlier: are we sure branching won't happen in the execution of the kernel?

**Review by mlxd — COMMENTED** [review]
Thanks @vincentmr 
I think we can come back and discuss this in a few weeks

File: pennylane_lightning/core/src/simulators/lightning_gpu/initSV.cu
Comment (vincentmr) [reply pennylane_lightning/core/src/simulators/lightning_gpu/initSV.cu:124]: Given the perceived urgency at the time, I simply mimicked the other L-GPU kernels without giving it much thought. That said, this is a common pattern in the CUDA [doc](https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#initialization). I think it is there to prevent out of bound access when the array is too small or when the array length is not divisible by the number of threads per block. So toward the end of the array you can't fill the SMs if the array is not commensurate with the number of SMs, but otherwise the branch should not prevent using the SMs fully.

File: pennylane_lightning/core/src/simulators/lightning_gpu/initSV.cu
Comment (vincentmr) [reply pennylane_lightning/core/src/simulators/lightning_gpu/initSV.cu:144]: I think that since the data (and access pattern) is intrinsically 1D, CUDA will in effect linearize any 2D/3D mapping. We could even avoid using `dim3` here (copy-pasting) and elsewhere. The `<<<numBlocks, threadsPerBlock>>>` syntax should be sufficient, but maybe we'd lose the possibility to call this asynchronously with streams?

File: pennylane_lightning/core/src/simulators/lightning_gpu/gates/tests/Test_StateVectorCudaManaged_Param.cpp
Comment (mlxd) [inline pennylane_lightning/core/src/simulators/lightning_gpu/gates/tests/Test_StateVectorCudaManaged_Param.cpp:39]: Do we want this defined in the translation unit, or local to the tests it is used in?

File: pennylane_lightning/core/src/simulators/lightning_gpu/initSV.cu
Comment (mlxd) [inline pennylane_lightning/core/src/simulators/lightning_gpu/initSV.cu:19]: Isn't there a CUDA function for this already defined in their API ?

File: pennylane_lightning/core/src/simulators/lightning_gpu/initSV.cu
Comment (mlxd) [inline pennylane_lightning/core/src/simulators/lightning_gpu/initSV.cu:26]: Same q for here?

File: pennylane_lightning/core/src/simulators/lightning_gpu/initSV.cu
Comment (mlxd) [inline pennylane_lightning/core/src/simulators/lightning_gpu/initSV.cu:286]: Not sure this file is the best overall for the location of these. Maybe we make an additional story to formalise the CUDA kernels when exploring the additions for the CPU-GPU box.

File: tests/test_gates.py
Comment (mlxd) [inline tests/test_gates.py:56]: Should we add a controlled version test here too?

**Review by mlxd — COMMENTED** [review]
Thanks @vincentmr 
Just a few suggestions, and things to follow up on but overall happy to approve for now when the comments are discussed

File: pennylane_lightning/core/src/simulators/lightning_gpu/utils/cuda_helpers.hpp
Comment (mlxd) [inline pennylane_lightning/core/src/simulators/lightning_gpu/utils/cuda_helpers.hpp:79]:
```suggestion
__host__ __device__ inline static constexpr auto Conj(CFP_t a) -> CFP_t {
```

Should we also make this host compatible?

File: pennylane_lightning/core/src/simulators/lightning_gpu/utils/cuda_helpers.hpp
Comment (mlxd) [inline pennylane_lightning/core/src/simulators/lightning_gpu/utils/cuda_helpers.hpp:98]:
```suggestion
__host__ __device__ inline static constexpr auto Cmul(CFP_t a, CFP_t b) -> CFP_t {
```

Same q as above?

**Review by mlxd — COMMENTED** [review]
Thanks @vincentmr 
Last 2 comments to add. Once the given are addressed, I can approve this. We can backlog the extension for adjoint support

File: pennylane_lightning/core/src/simulators/lightning_gpu/utils/cuda_helpers.hpp
Comment (vincentmr) [reply pennylane_lightning/core/src/simulators/lightning_gpu/utils/cuda_helpers.hpp:79]: It seems it is not required at present, but we might as well.

File: tests/test_gates.py
Comment (mlxd) [reply tests/test_gates.py:56]: Sounds good to me then. Thanks

File: pennylane_lightning/core/src/simulators/lightning_gpu/gates/tests/Test_StateVectorCudaManaged_Param.cpp
Comment (mlxd) [reply pennylane_lightning/core/src/simulators/lightning_gpu/gates/tests/Test_StateVectorCudaManaged_Param.cpp:39]: Yes, that exactly. I know this has caused problems in the Python layer in the past.

File: pennylane_lightning/core/src/simulators/lightning_gpu/StateVectorCudaManaged.hpp
Comment (mlxd) [reply pennylane_lightning/core/src/simulators/lightning_gpu/StateVectorCudaManaged.hpp:222]: Since this is currently working, we can accept as is, and aim to validate later with different tuning.

**Comment by vincentmr** [comment]
Lightning adjoint diff supports GlobalPhase [sc-56497]

File: pennylane_lightning/core/src/simulators/lightning_gpu/gates/tests/Test_StateVectorCudaManaged_Param.cpp
Comment (vincentmr) [reply pennylane_lightning/core/src/simulators/lightning_gpu/gates/tests/Test_StateVectorCudaManaged_Param.cpp:39]: Ok, I'll fix this except for GPU-MPI which I'm not testing in this PR.

File: pennylane_lightning/core/src/simulators/lightning_gpu/initSV.cu
Comment (multiphaseCFD) [reply pennylane_lightning/core/src/simulators/lightning_gpu/initSV.cu:140]: Thanks @mlxd! Good question. Threads number per block is default as `256`, which is always less than the number of allowed threads per block as of `Ampere` card. `256` threads per block was chosen (which is mentioned in the following comment) is because this is the minimal number of threads per block (8 warps) required to hide maximal latency.

File: pennylane_lightning/core/_serialize.py
Comment (AmintorDusko) [inline pennylane_lightning/core/_serialize.py:411]: This doesn't look like a serialize-related functionality. Let's look for a better place since this file is designed to host the `QuantumScriptSerializer` class.

File: .readthedocs.yml
Comment (mlxd) [inline .readthedocs.yml:31]: Nice work on this.

File: pennylane_lightning/core/src/simulators/lightning_gpu/gates/tests/Test_StateVectorCudaManaged_Param.cpp
Comment (AmintorDusko) [inline pennylane_lightning/core/src/simulators/lightning_gpu/gates/tests/Test_StateVectorCudaManaged_Param.cpp:1614]:
```suggestion
```

File: pennylane_lightning/core/src/simulators/lightning_qubit/gates/tests/Test_GateImplementations_Param.cpp
Comment (AmintorDusko) [inline pennylane_lightning/core/src/simulators/lightning_qubit/gates/tests/Test_GateImplementations_Param.cpp:2700]:
```suggestion
```

File: pennylane_lightning/core/src/gates/Gates.hpp
Comment (mlxd) [inline pennylane_lightning/core/src/gates/Gates.hpp:124]: Wait, was this incorrect in the previous definition?

File: pennylane_lightning/core/src/simulators/lightning_gpu/gates/tests/Test_StateVectorCudaManaged_Param.cpp
Comment (mlxd) [inline pennylane_lightning/core/src/simulators/lightning_gpu/gates/tests/Test_StateVectorCudaManaged_Param.cpp:1614]:
```suggestion
```

File: pennylane_lightning/lightning_gpu/lightning_gpu.py
Comment (AmintorDusko) [inline pennylane_lightning/lightning_gpu/lightning_gpu.py:541]: The `global_phase_diagonal` says that it returns only the diagonal of the operator.
Are we only sending the diagonal to be applied? I got confused here.

File: pennylane_lightning/core/src/simulators/lightning_kokkos/gates/tests/Test_StateVectorKokkos_Param.cpp
Comment (mlxd) [inline pennylane_lightning/core/src/simulators/lightning_kokkos/gates/tests/Test_StateVectorKokkos_Param.cpp:250]: Any reason for the order swap?

**Review by AmintorDusko — COMMENTED** [review]
Hi @vincentmr, I left a few comments.

File: pennylane_lightning/core/src/simulators/lightning_qubit/gates/cpu_kernels/GateImplementationsLM.hpp
Comment (mlxd) [inline pennylane_lightning/core/src/simulators/lightning_qubit/gates/cpu_kernels/GateImplementationsLM.hpp:1867]: Is calling find in a loop like this ok for the usage we expect?

File: pennylane_lightning/core/src/utils/TestHelpersWires.hpp
Comment (mlxd) [inline pennylane_lightning/core/src/utils/TestHelpersWires.hpp:64]: Is this block needed?

File: requirements-dev.txt
Comment (mlxd) [inline requirements-dev.txt:3]: Is this not a built in library?

File: tests/test_gates.py
Comment (mlxd) [inline tests/test_gates.py:56]: Should we add an out of order wire test too?

**Review by mlxd — COMMENTED** [review]
Thanks @vincentmr 
Just a few final questions, and will be happy to approve once they are commented on.

File: pennylane_lightning/core/src/simulators/lightning_gpu/gates/tests/Test_StateVectorCudaManaged_Param.cpp
Comment (vincentmr) [reply pennylane_lightning/core/src/simulators/lightning_gpu/gates/tests/Test_StateVectorCudaManaged_Param.cpp:1614]: This is a comment saying what the `phase` diagonal is and how is it generate. I would like to keep it in case we want to change it in the future.

File: pennylane_lightning/core/_serialize.py
Comment (vincentmr) [reply pennylane_lightning/core/_serialize.py:411]: This functions takes in a [pennylane.operation.Operation](https://docs.pennylane.ai/en/stable/code/api/pennylane.operation.Operation.html) and spits out a NumPy array to be consumed by the C++ layer. It is a serialization util in that sense. Where do you see this go otherwise?

File: pennylane_lightning/core/src/simulators/lightning_qubit/gates/tests/Test_GateImplementations_Param.cpp
Comment (vincentmr) [reply pennylane_lightning/core/src/simulators/lightning_qubit/gates/tests/Test_GateImplementations_Param.cpp:2700]: [Same](https://github.com/PennyLaneAI/pennylane-lightning/pull/579#discussion_r1491305967).

File: pennylane_lightning/lightning_gpu/lightning_gpu.py
Comment (vincentmr) [reply pennylane_lightning/lightning_gpu/lightning_gpu.py:541]: Yes, the diagonal is passed via the `matrix` argument down to `applyOperation` which uses a special routine when the name of the operation is `C(GlobalPhase)`. I used that hack to get something working quick in L-GPU/Kokkos. The proper way to do things will be to add controlled-kernels as in L-Qubit, after which we can delete `global_phase_diagonal` and the associated logic.

File: pennylane_lightning/core/src/gates/Gates.hpp
Comment (vincentmr) [reply pennylane_lightning/core/src/gates/Gates.hpp:124]: `SQRT2<decltype(ONE<ComplexT, T>().x)>() / 2` is correct but a roundabout way to get `INVSQRT2<ComplexT, T>()`. 
`-IMAG<ComplexT, T>()` was a mistake. During the controlled-gate implementation I found many other such [typos] in that module which isn't tested much it seems.

File: pennylane_lightning/core/src/simulators/lightning_kokkos/gates/tests/Test_StateVectorKokkos_Param.cpp
Comment (vincentmr) [reply pennylane_lightning/core/src/simulators/lightning_kokkos/gates/tests/Test_StateVectorKokkos_Param.cpp:250]: Just seemed more natural to iterate from 0 to 1 (casting in my head), or start with the default inverse value.

File: pennylane_lightning/core/src/simulators/lightning_qubit/gates/cpu_kernels/GateImplementationsLM.hpp
Comment (vincentmr) [reply pennylane_lightning/core/src/simulators/lightning_qubit/gates/cpu_kernels/GateImplementationsLM.hpp:1867]: I didn't like that piece of code either, but it should not impact performance in any case. The problem is the following: 

- `controlled_wires` is a small array with length at most `num_qubits - 1`.
- we want a wire, any wire, that's not in `controlled_wires` and set the wire for the kernel to that.

Can you think of a nice 1-/2-liner to get this on top of your head?

File: pennylane_lightning/core/src/utils/TestHelpersWires.hpp
Comment (vincentmr) [reply pennylane_lightning/core/src/utils/TestHelpersWires.hpp:64]: This is preparation for #614 . I would like to uncomment it once that gets merged.

File: requirements-dev.txt
Comment (vincentmr) [reply requirements-dev.txt:3]: Yes, but I was getting an error related to `datetime` in the CI so I tried upgrading it. Probably one of those CI flukes, let me try to remove it now.

File: tests/test_gates.py
Comment (vincentmr) [reply tests/test_gates.py:56]: I could, but then should we do it for all the other gates? This is probably the least sensitive gate since the `wires` argument is actually unused in the C++ kernels.

File: pennylane_lightning/core/src/simulators/lightning_qubit/gates/cpu_kernels/GateImplementationsLM.hpp
Comment (mlxd) [reply pennylane_lightning/core/src/simulators/lightning_qubit/gates/cpu_kernels/GateImplementationsLM.hpp:1867]: Based on that, I think for now I think this is fine. If the order was sorted, it'd be faster to do bin-search, but while this isn't a bottleneck let's go for it.

File: requirements-dev.txt
Comment (mlxd) [reply requirements-dev.txt]: Weird. Fine to keep if needed for the CI.

File: tests/test_gates.py
Comment (mlxd) [reply tests/test_gates.py:56]: Let's look at this in a follow-up improvement of the test suite.

**Review by mlxd — APPROVED** [review]

File: pennylane_lightning/core/src/simulators/lightning_gpu/initSV.cu
Comment (AmintorDusko) [inline pennylane_lightning/core/src/simulators/lightning_gpu/initSV.cu:274]: Can we do that?
```suggestion
void cGlobalPhaseStateVector_CUDA(cuComplex *sv, size_t num_sv, const bool adjoint,
                                  cuComplex *phase, size_t thread_per_block,
                                  cudaStream_t stream_id) {
    cGlobalPhaseStateVector_CUDA_call<cuComplex, size_t, adjoint>(
            sv, num_sv, phase, thread_per_block, stream_id);
}
void cGlobalPhaseStateVector_CUDA(cuDoubleComplex *sv, size_t num_sv,
                                  const bool adjoint, cuDoubleComplex *phase,
                                  size_t thread_per_block,
                                  cudaStream_t stream_id) {
    cGlobalPhaseStateVector_CUDA_call<cuDoubleComplex, size_t, adjoint>(
            sv, num_sv, phase, thread_per_block, stream_id);
}
```

File: pennylane_lightning/core/src/simulators/lightning_kokkos/StateVectorKokkos.hpp
Comment (AmintorDusko) [inline pennylane_lightning/core/src/simulators/lightning_kokkos/StateVectorKokkos.hpp:731]:
```suggestion
    void applyGlobalPhase(const std::vector<size_t> &wires,
                          const bool inverse = false,
                          const std::vector<fp_t> &params = {}) {
        auto &&num_qubits = this->getNumQubits();

       Kokkos::parallel_for(
                Kokkos::RangePolicy<KokkosExecSpace>(0, exp2(num_qubits)),
                globalPhaseFunctor<fp_t, inverse>(*data_, num_qubits, wires,
                                               params));
    }
```

File: pennylane_lightning/lightning_gpu/lightning_gpu.py
Comment (AmintorDusko) [reply pennylane_lightning/lightning_gpu/lightning_gpu.py:541]: OK

File: pennylane_lightning/core/src/simulators/lightning_gpu/initSV.cu
Comment (vincentmr) [reply pennylane_lightning/core/src/simulators/lightning_gpu/initSV.cu:274]: No because `adjoint` needs to be a comptime constant.

File: pennylane_lightning/core/src/simulators/lightning_kokkos/StateVectorKokkos.hpp
Comment (vincentmr) [reply pennylane_lightning/core/src/simulators/lightning_kokkos/StateVectorKokkos.hpp:731]: Same, we can't use a variable (even `const`) in a template parameter.

File: pennylane_lightning/core/_serialize.py
Comment (vincentmr) [reply pennylane_lightning/core/_serialize.py:411]: I would leave it here for the time being, but I don't mind moving it to `lightning_base.py` if you prefer. But I don't think we should create a module solely for that.

Comment (AmintorDusko) [reply pennylane_lightning/core/_serialize.py:411]: OK. We can come back to this in the future.

File: pennylane_lightning/core/src/simulators/lightning_kokkos/StateVectorKokkos.hpp
Comment (AmintorDusko) [reply pennylane_lightning/core/src/simulators/lightning_kokkos/StateVectorKokkos.hpp:731]: Are you sure? Because I tested it with a std::array and it worked. :laughing:

Comment (AmintorDusko) [reply pennylane_lightning/core/src/simulators/lightning_kokkos/StateVectorKokkos.hpp:731]: Anyway, you don't need to spend time on that, we can leave it like this for now.

Comment (vincentmr) [reply pennylane_lightning/core/src/simulators/lightning_kokkos/StateVectorKokkos.hpp:731]: This does not work on my side, but you're right std::array templates are somehow more flexible.

**Pull request merged**
